### DISCUSSION TITLE
XGBoostLSS Feature Requests

### DISCUSSION THREAD

**Discussion by StatMixedML** [StatMixedML/XGBoostLSS#24] [Ideas]
Dear Community,

In case you find the XGBoostLSS library interesting for your research / use-case, please use this forum for posting new feature requests you would like to see in the future.

I am currently working on a more efficient package structure. So stay tuned.

**Comment by kmedved** [comment]
Interesting! Can you provide more details?

Also, is there any chance of getting an sklearn compatible API?

**Reply by StatMixedML** [reply]
@kmedved Thanks for your comment!

In its current version, there are potential problems when it comes to serialization of the models. So I am planning to have a better class structure of the package. Also, I fully base the package on PyTorch functions for the derivatives that make it easier to generalize. 

Currently, I am not planning to have a sklearn integration. What advantages do you see in that?

**Reply by kmedved** [reply]
The main advantage of sklearn is just integration with the rest of the python DS ecosystem. It's very convenient to have a model that you can use as a drop-in replacement for testing purposes, and to interact with the other parts of the scikit-learn ecosystem.

Either way, I appreciate all your work on this package, but just a suggestion to help with adoption. Thanks!

**Reply by StatMixedML** [reply]
I appreciate your suggestion, don't get me wrong. It is just that embedding it into the sklearn framework creates a lot of outside dependencies, which I want to avoid.

**Comment by HiramCillo** [comment]
I want to try the multi-target regression, because I want to try it on a
task called " match_modality" for single-cell multiomics

**Reply by StatMixedML** [reply]
@HiramCillo Thanks for your comment!

Right now, the multivariate version can be slow to estimate for high-dimensional outputs. I am also currently working on adding a new boosting model to the repo that is faster. 
 
Of what dimension are we talking in terms of number of targets for the multi-output regression?

**Comment by StatMixedML** [comment]
What extensions (additional distributions, runtime, etc.) of XGBoostLSS do you consider important to make it more competitive in comparison to other approaches?

**Reply by joseortiz3** [reply]
I'm wondering if the natural gradient approach (e.g. in NGBoost) can be incorporated into XGBoostLSS? It might improve convergence and stability, provided the hessians corresponding to the natural gradients can be analytically found. Seems like putting natural gradients into XGBoostLSS would certainly be faster/more optimized than NGBoost since XGBoost would be the underlying framework.

**Reply by StatMixedML** [reply]
Thanks for your interest in the project!

I am not exactly sure how NGBoost is working. As far as XGBoostLSS is concerned, it trains all parameters of distribution, using a Newton-Raphson type of algorithm, i.e., a second order Taylor approximation of the loss function

<p align="center">

<img src="https://user-images.githubusercontent.com/41187941/232194697-ad21c5cc-7691-4a8d-a0be-493871886c67.png" width="400" height="60">

</p>

where for each distributional parameter $\theta_{m}$, gradients and hessians are calculated

<p align="center">

<img src="https://user-images.githubusercontent.com/41187941/232194834-4a454915-8adb-4e9d-9a03-311c33050ca5.png" width="300" height="120">

</p>

Since gradients and hessians do not have to be analytically derived, we can leverage automatic differentiation to derive them numerically. This allows to also use complex loss functions, such as Normalizing Flows as shown in [Distributional Gradient Boosting Machines](https://arxiv.org/abs/2204.00778) or multivariate distributions as show in [Multi-Target XGBoostLSS Regression](https://arxiv.org/abs/2210.06831). 

You also mention that

> [...] It might improve convergence and stability [...]

How would natural gradients (including hessians) be different from the XGBoostLSS gradients and hessians?

**Reply by joseortiz3** [reply]
> How would natural gradients (including hessians) be different from the XGBoostLSS gradients and hessians?

Natural gradients differ from normal gradients in that the "best direction" for loss improvement is invariant to a user's choice of how they parameterize a given distribution. It uses a different notion of distance, one that is better suited for probability distributions (e.g. KL divergence instead of euclidian). From the [NGBoost paper](https://arxiv.org/pdf/1910.03225.pdf):

<img width="965" alt="image" src="https://user-images.githubusercontent.com/18585219/232252009-1bf27bab-2e62-4d36-9128-c79bac2cce5d.png">

<img width="346" alt="image" src="https://user-images.githubusercontent.com/18585219/232252100-f95a5bbe-1724-49f5-b382-3aa9c8e1e779.png">

The paper's algorithm is a standard psuedo-residuals-fitting (gradient-fitting) booster, but it doesn't yet leverage hessians like XGBoost. They've worked out the natural gradients analytically for [many different distributions](https://github.com/stanfordmlgroup/ngboost/tree/master/ngboost/distns) explained [here](https://stanfordmlgroup.github.io/ngboost/5-dev.html). I believe adopting the NGBoost natural gradients in XGBoostLSS parameter updates could be the best of both worlds.

<img width="350" alt="image" src="https://user-images.githubusercontent.com/18585219/232252273-37a98cb9-c106-472d-b4c4-08933f408133.png">

**Reply by StatMixedML** [reply]
After skimming through the paper briefly, it says

<p align="center">

<img src="https://user-images.githubusercontent.com/41187941/232283910-1bb15d55-43f5-42de-8284-7833ba73ba36.png" width="350" height="500">

</p>

From (8) we see that using MLE and the log-likelihood as the optimization metric, the natural gradient is given by the product of the inverse of the Fisher information matrix (which is a matrix that describes the curvature of the objective function with respect to the model parameters) and the usual gradient. Hence, scaling the usual gradient by the inverse fisher information matrix can provide a more accurate and efficient update direction for the model parameters in certain scenarios where the curvature of the objective function is significant, such as when the parameters are highly correlated, or the parameter space is highly curved.



Since XGBoostLSS currently optimizes on the log-likelihood only (while NGBoost also allows to optimize on scoring rules such as CRPS), one could try scaling the gradients in XGBoostLSS with the inverse Hessian (given the that the Hessian is invertible), since the negative Hessian evaluated at the MLE is the same as the observed Fisher information matrix evaluated at the MLE. However, due to efficiency reasons, XGBoostLSS currently only uses the diagonal of the Hessian matrix, and hence, disregards the cross-derivatives between distributional parameters. I am not sure if NGBoost makes use of the full Hessian matrix, also using the off-diagonal elements. Anyone knows this? 

Also, it is important to note that inverting a Hessian might easily lead to numerical instabilities, especially for high-dimensional parameter settings, or when parameters are co-dependent. In extreme cases, this can cause the inverse even to be undefined, resulting in inaccurate or unreliable results. One way to mitigate this would be to add a small positive value to the diagonal elements of the Hessian to improve numerical stability during its inversion. Anyone knows how NGBoost handles such cases?
